MapboxNavigation/InstructionPresenter.swift

```diff
@@ -0,0 +1,87 @@
+import UIKit
+import MapboxDirections
+
+class InstructionPresenter {
+    var instruction: [VisualInstructionComponent]?
+
+    typealias ShieldDownloadCompletion = (NSAttributedString) -> ()
+    var onShieldDownload: ShieldDownloadCompletion?
+
+    func attributedTextForLabel(_ label: InstructionLabel) -> NSAttributedString {
+        guard let instruction = instruction else {
+            return NSAttributedString()
+        }
+
+        let string = NSMutableAttributedString()
+
+        // Add text or image
+        for component in instruction {
+            let isFirst = component == instruction.first
+            let joinChar = !isFirst ? " " : ""
+
+            if let shieldKey = component.shieldKey() {
+                if let cachedImage = component.cachedShield(shieldKey) {
+                    string.append(attributedString(withFont: label.font, shieldImage: cachedImage))
+                } else {
+                    // Download shield and display road code in the meantime
+                    if let text = component.text {
+                        string.append(NSAttributedString(string: joinChar + text, attributes: attributesForLabel(label)))
+                    }
+                    component.shieldImage(height: label.shieldHeight, completion: { [unowned self] (image) in
+                        guard image != nil, component.cachedShield(shieldKey) != nil else {
+                            return
+                        }
+                        if let completion = self.onShieldDownload {
+                            completion(self.attributedTextForLabel(label))
+                        }
+                    })
+                }
+            } else if let text = component.text {
+                if component.type == .delimiter && instructionHasDownloadedAllShields() {
+                    continue
+                }
+                string.append(NSAttributedString(string: (joinChar + text).abbreviated(toFit: label.availableBounds(), font: label.font), attributes: attributesForLabel(label)))
+            }
+        }
+
+        return string
+    }
+
+    private func instructionHasDownloadedAllShields() -> Bool {
+        for component in instruction! {
```

Comment:
Nit: this seems to be safe considering the guard in `attributedTextForLabel(_:)` but the linter will complain about unwrapping when we decide to enable it.